### ISSUE TITLE
[Feature] Add LatentMPCController — caller-owned CEM/MPC receding-horizon planning over the `score` capability

### ISSUE THREAD

**Issue by AbdelStark** [AbdelStark/worldforge#328]
## Summary

Add a first-class, caller-owned **`LatentMPCController`** — a Model-Predictive-Control planner that wraps the `score` (Cost) capability in a real sample → roll-out → score → refit → execute-first-action → replan loop — and **wire it through `World.plan(planner="latent-mpc", ...)`** as a new planning mode alongside the existing `policy+score` / `policy` / `score` / `predict` paths. `LatentMPCController` is the reusable engine; `World.plan` is the public, ergonomic entry. This realizes the currently-unused `plan` capability flag as a provider-agnostic optimizer and lets the robotics showcase depict how latent world models are *actually* used for control, instead of the present single-shot pick over fabricated candidates.

## Motivation

The PushT showcase ([`scripts/robotics-showcase`](../../scripts/robotics-showcase), [`src/worldforge/smoke/lerobot_leworldmodel.py`](../../src/worldforge/smoke/lerobot_leworldmodel.py)) currently runs the `policy+score` path in [`framework.py`](../../src/worldforge/framework.py) (`_plan_with_policy_and_score`): the LeRobot policy produces **one** action chunk, [`smoke/pusht_showcase_inputs.py`](../../src/worldforge/smoke/pusht_showcase_inputs.py) fabricates three arithmetic variants of it (`base`, `0.5*base`, `-0.5*base`), `LeWorldModelProvider.score_actions` scores the three, and the planner takes `argmin`. There is **no proposal distribution, no iterative refinement, no receding horizon**, and the goal frame comes from an env reset (seed 11) unrelated to the observation reset (seed 7).

This validates the integration *plumbing* but does not correspond to any recognized control paradigm. The published latent world models WorldForge targets use a different, well-defined procedure:

- **V-JEPA 2-AC** (arXiv:2506.09985): test-time **MPC + Cross-Entropy Method (CEM)**; minimizes `E = ||P(a^{1:T}; s_k, z_k) - z_g||_1` (L1 latent distance to an image goal); **execute the first action, then replan**; 800 samples × 10 refinement iters per step.
- **LeWM** (arXiv:2603.19312): same structure — **CEM/MPC** in latent space, cost `C = ||z_hat_H - z_g||_2^2`, execute first K actions then replan; <1 s/plan.

In both, the world model is a **cost oracle that stays in the control loop**; the optimizer (CEM/MPPI) lives in the caller. WorldForge already exposes the right primitive — `ActionScoreResult` is a batched cost over candidates — but is missing the loop that turns that cost into a controller. Note that `stable-worldmodel` itself ships CEM/MPPI/Gradient-Descent/Random planners (per [`specs/roboto-showcase/spec.md`](../../specs/roboto-showcase/spec.md)); a WorldForge-native controller gives us the same capability **provider-agnostically**, over any `score` provider (`leworldmodel`, future `jepa-wms` per #325, future arbitration per #326), without binding to one runtime's planner.

## Goal

Public entry: **`World.plan(planner="latent-mpc", ...)`** runs **one MPC solve** (the CEM refinement loop) over horizon `H` and returns a `Plan` whose `actions` are the optimized action chunk — exactly like the other planning modes return a `Plan`. The closed-loop receding horizon is achieved by the **host re-calling `World.plan` after executing `execute_k` actions and re-observing** (env stepping is host/sim-owned, consistent with the optional-runtime boundary). One MPC solve:

```
for iteration in range(num_iterations):
    candidates = sampler.sample(mu, sigma, num_samples, horizon)   # action space, pure Python
    payload    = encoder.encode(candidates, observation_info, goal_info)  # action space -> provider score payload
    result     = world.score(score_provider, info=payload, action_candidates=...)
    elites     = topk(candidates, result.scores, num_elites, lower_is_better=result.lower_is_better)
    mu, sigma  = refit(elites)
return best elite (first execute_k actions) as a Plan, with per-iteration diagnostics in the trace metadata
```

`LatentMPCController` is the engine that performs this solve (`plan_step() -> MPCStepResult`); `World._plan_with_mpc` adapts its result into a `Plan`.

## Non-goals

- **No training.** This is an inference-time controller; world-model/policy training stays in upstream repos (consistent with the optional-runtime boundary).
- **No new base dependencies.** The CEM/MPPI math must run on the existing base stack (stdlib `random`/`statistics`; `numpy` only if it is already acceptable — confirm). All tensor work stays inside the provider's lazily-imported runtime. **Do not add `torch` to `src/worldforge/` base imports.**
- **Not removing** the existing single-shot `policy+score` path; this adds a new, honest planning mode alongside it.
- **No real-hardware loop, no robot safety layer** (out of scope per repo invariants).

### Module placement
The optimizer engine lives in a new module to keep [`framework.py`](../../src/worldforge/framework.py) lean: `src/worldforge/control/` (or `src/worldforge/planning/`) — `mpc.py` with `LatentMPCController` plus the sampler/encoder protocols and result types. `framework.py` imports it for the dispatch adapter. This keeps the CEM loop out of the framework while making `World.plan` the public entry.

### `World.plan` contract extension (GATED — requires approval; granted by repo owner in this issue)
Modifying [`framework.py`](../../src/worldforge/framework.py) planning behavior is a **gated public-contract change** per repo invariants. The extension is deliberately additive and backward compatible:

- **Dispatch:** add `_PlanRequest.uses_mpc_planning` (derived in `_build_plan_request` from `planner == "latent-mpc"`) and check it **first** in `plan(...)`:
  ```python
  if request.uses_mpc_planning:
      return self._plan_with_mpc(request, resolved_goal)
  if request.uses_policy_planning:
      return self._plan_with_policy(request, resolved_goal)
  if request.uses_score_planning:
      return self._plan_with_score(request, resolved_goal)
  return self._plan_with_predictions(request, resolved_goal)
  ```
- **New `_plan_with_mpc(request, resolved_goal) -> Plan`** sits beside `_plan_with_policy_and_score`; it constructs/invokes `LatentMPCController`, runs one MPC solve, and adapts `MPCStepResult` → `Plan` (chosen `execute_k` actions). It requires a `score_provider` (the cost oracle) and optionally a `policy_provider` (warm-start); raises `WorldForgeError` if `score_provider` is absent.
- **New optional `World.plan(...)` keywords** (all default `None`, keyword-only, backward compatible):
  - `planner_config: PlannerConfig | None` — horizon / samples / iterations / elites / execute_k / init_std / seed / proposal.
  - `candidate_encoder: ScoreCandidateEncoder | None` — action-space candidates → `score` payload (provider/bridge-supplied).
  - `goal_info: JSONDict | None` — goal payload; `score_info` carries the current observation payload the encoder augments per iteration.
- **Return contract preserved:** `_plan_with_mpc` returns a `Plan` (one solve = open-loop optimized chunk). Per-iteration CEM diagnostics (`iteration_costs`, candidate counts, latencies) go into the workflow-trace metadata next to the existing `"planning_mode": "mpc"` tag (mirrors the `policy+score` trace at framework.py ~1611). **Receding horizon is the host re-calling `World.plan`** — no env I/O enters the planning contract.
- Add `_PlanRequest` fields for `planner_config` / `candidate_encoder` / `goal_info`; extend `_build_plan_request` accordingly.

### Key architectural constraint — keep the loop in action space, torch-free
The controller samples and refits **in WorldForge `Action` space** (`list[list[Action]]` = N candidate sequences of H actions), operating only on the numeric action fields. A per-provider **encoder** converts action-space candidates into the provider's `score` payload (e.g. PushT's `(1, N, H, 10)` tensor + `pixels`/`goal`/`action` info), so all tensor/torch work stays inside the lazily-imported provider runtime. This is what makes the controller provider-agnostic and base-dependency-clean.

### Proposed surfaces (illustrative)

```python
@dataclass(slots=True, frozen=True)
class PlannerConfig:
    horizon: int                  # H: action-sequence length rolled out by the world model
    num_samples: int              # N: candidates per CEM iteration (e.g. 256-512)
    num_iterations: int           # M: CEM refinement iterations (e.g. 5-10)
    num_elites: int               # top-k kept to refit the proposal distribution
    execute_k: int = 1            # receding horizon: actions executed before replanning
    init_std: float = 1.0
    seed: int | None = None       # determinism
    proposal: Literal["gaussian", "policy"] = "gaussian"

class CandidateSampler(Protocol):
    def sample(self, mean, std, num_samples, horizon) -> list[list[Action]]: ...

class ScoreCandidateEncoder(Protocol):
    """Provider/bridge-supplied: action-space candidates -> score_actions payload (info, action_candidates)."""
    def encode(self, candidates: list[list[Action]], *, observation_info, goal_info) -> tuple[JSONDict, object]: ...

@dataclass(slots=True, frozen=True)
class MPCStepResult:
    actions: list[Action]                 # chosen chunk (first execute_k after the step)
    best_cost: float
    candidate_count: int
    iteration_costs: list[float]          # best-elite cost per iteration (must be non-increasing on convex toy)
    metadata: JSONDict                    # latencies, lower_is_better, provider, planner label

class LatentMPCController:
    def __init__(self, *, world: World, score_provider: str,
                 encoder: ScoreCandidateEncoder,
                 config: PlannerConfig,
                 sampler: CandidateSampler | None = None,
                 policy_provider: str | None = None) -> None: ...
    def plan_step(self, *, observation_info: JSONDict, goal_info: JSONDict) -> MPCStepResult: ...
```

- **Composition A (`proposal="gaussian"`)** — pure world-model MPC (the V-JEPA 2-AC / LeWM depiction): the controller is the policy; no `policy_provider` needed.
- **Composition B (`proposal="policy"`)** — policy warm-start (TD-MPC2 / Dream-MPC style): seed `mu` (and some samples) from `policy_provider.select_actions(...).action_candidates` (ideally a diffusion policy that emits genuine multi-sample proposals), then refine with the world-model cost. This is the *legitimate* version of "policy proposes, world model scores" — real samples + iterative refinement, not arithmetic perturbations.

### Validation & error contract
- Validate `PlannerConfig` (`0 < num_elites <= num_samples`, `horizon >= 1`, `num_iterations >= 1`, `execute_k <= horizon`) at construction; raise `WorldForgeError` on invalid public inputs.
- After each `score_actions`, assert `len(scores) == num_samples`; raise `WorldForgeError` on mismatch (mirror existing `_score_candidate_action_plans`).
- Respect `ActionScoreResult.lower_is_better` for elite selection (don't hardcode argmin).

## Showcase changes

Update [`src/worldforge/smoke/lerobot_leworldmodel.py`](../../src/worldforge/smoke/lerobot_leworldmodel.py):
- Add an MPC mode that calls `World.plan(planner="latent-mpc", score_provider="leworldmodel", planner_config=..., candidate_encoder=..., score_info=obs, goal_info=goal)`; provide a PushT `ScoreCandidateEncoder` (refactor [`smoke/pusht_showcase_inputs.py`](../../src/worldforge/smoke/pusht_showcase_inputs.py) so `build_score_info` + action-block lifting become the encoder, and **delete the `{base, 0.5*base, -0.5*base}` fabrication**).
- Construct the goal from a **coherent task state** (goal pose / goal image on the same trajectory), not an unrelated env seed.
- Run the host-owned receding-horizon loop: execute `execute_k` actions against the PushT env, re-observe, and **re-call `World.plan(planner="latent-mpc", ...)`**; emit JSON + run-manifest with the new diagnostics (`iteration_costs`, steps-to-goal, planning latency).
- Keep the existing single-shot `policy+score` path available for comparison.

## Testing plan
- **Deterministic convex toy** `score` provider (cost = distance of action to a fixed known optimum, `lower_is_better=True`), seeded RNG: assert `iteration_costs` is non-increasing and converges near the optimum; assert chosen action approaches the optimum as `num_iterations`/`num_samples` grow. No torch, fully deterministic.
- Config-validation failures raise `WorldForgeError`.
- Elite selection honors `lower_is_better` (test both directions).
- Composition B: a fake `policy` provider seeds `mu`; assert it converges in fewer iterations than Composition A on the toy.
- Score-length mismatch raises `WorldForgeError`.
- Keep coverage at/above the 90% gate; controller core must be exercised without optional runtimes.

## Docs
- New planning/control concept page (oracle vs controller vs planner); document Compositions A/B and when to use each (latency/data budget).
- Clarify in `AGENTS.md` / docs: `score` = batched cost oracle; `policy` = controller/proposal; pairing them is meaningful only as MPC (Composition B), not "perturb-and-pick".
- Update the showcase/provider docs and `CHANGELOG.md` for the new planning mode.

## Acceptance criteria
- [ ] `LatentMPCController` + `PlannerConfig` + sampler/encoder protocols + `MPCStepResult` land in a new `src/worldforge/control/` (or `planning/`) module; **no `torch` in base imports**.
- [ ] CEM loop runs entirely in action space; provider tensor work stays behind the lazy runtime import.
- [ ] **`World.plan(planner="latent-mpc", ...)` works:** `_PlanRequest.uses_mpc_planning` + `_plan_with_mpc` dispatch added; new `planner_config` / `candidate_encoder` / `goal_info` keywords are optional and backward compatible; returns a `Plan`; missing `score_provider` raises `WorldForgeError`; `"planning_mode": "mpc"` + `iteration_costs` recorded in trace metadata. Existing `policy+score`/`policy`/`score`/`predict` paths unchanged (regression-tested).
- [ ] Composition A and B both work via `World.plan`; `lower_is_better` respected; invalid inputs raise `WorldForgeError`.
- [ ] PushT showcase has an MPC mode calling `World.plan(planner="latent-mpc", ...)`; fabricated `{base,0.5,-0.5}` candidates removed; goal is coherent; receding horizon closed by re-calling `World.plan`; JSON + run-manifest emit `iteration_costs` and steps-to-goal.
- [ ] Deterministic convex-toy tests prove monotone cost descent and convergence (both at the engine level and through `World.plan`); coverage gate >= 90% holds.
- [ ] Docs + AGENTS + CHANGELOG updated; `scripts/generate_provider_docs.py --check` clean if capability docs change.

## Open questions
- **Mode selector:** select MPC via the existing `planner` string (`planner="latent-mpc"`, as specified above) or introduce an explicit `planning_mode="mpc"` keyword? (Leaning `planner="latent-mpc"` since it reuses an existing arg and keeps the signature stable.) — *Resolved direction: wire through `World.plan`; the engine stays reusable in `src/worldforge/control/`.*
- **`numpy` in the controller:** acceptable in base, or stay stdlib-only? Affects sampler/refit implementation.
- **Optimizer family:** start with CEM only, or also expose MPPI / gradient-based behind the `proposal`/optimizer seam?
- **Goal-encoding contract:** standardize an image/state-goal contract across world-model providers so the controller is fully provider-agnostic (relates to #325, #326)?
- **Sub-goal chaining:** accept a sub-goal sequence (V-JEPA 2-AC uses sub-goal images) to mitigate the short-horizon limitation both papers acknowledge?

## References
- V-JEPA 2 / V-JEPA 2-AC — arXiv:2506.09985 (frozen ViT-g encoder; ~300M action-conditioned predictor; MPC+CEM; L1 latent-to-goal; ~16 s/action; Cosmos baseline ~4 min/action).
- LeWM — arXiv:2603.19312 (~15M end-to-end; `L = L_pred + λ·SIGReg`; CEM/MPC; L2 latent-to-goal; <1 s/plan).
- Background: Dreamer v3 (imagination-RL, model discarded at test time), TD-MPC2 / Dream-MPC (policy-prior + MPC), PETS (CEM over learned dynamics).
- Related: #326 (multi-world-model score arbitration), #325 (jepa-wms provider), #321 / `specs/roboto-showcase` (lifecycle showcase).

**Comment by omarespejel** [comment]
Upstream PR #331 is now the implementation candidate for this issue.

Current status:
- fork preflight: omarespejel/worldforge#31;
- upstream PR: #331;
- official CI: green across tests, quality, package contract, security, docs, and robotics showcase;
- local validation rerun today: `uv run pytest -q` -> `1720 passed, 92 skipped`, plus ruff, lock check, optional import boundary audit, docs snippets, mkdocs strict, and `git diff --check`.

The implementation keeps the boundary from this issue: pure-Python CEM/MPC over `score_actions(...)`; host/runtime code still owns simulator stepping, robot execution, optional torch/LeRobot/LeWorldModel dependencies, and embodiment-specific action translation.
